### PR TITLE
MySQL 8.4 Platform

### PR DESCRIPTION
**Pull request by wmouwen**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| Fixed issues | none

#### Summary

MySQL 8.4 introduces a new set of reserved keywords ([link][8.4-reserved-keywords]). This PR adds a new platform definition for MySQL 8.4 and higher.

[8.4-reserved-keywords]: https://dev.mysql.com/doc/refman/8.4/en/keywords.html#keywords-new-in-current-series

#### Related
- #6386

### CODE REVIEW

File: src/Driver/AbstractMySQLDriver.php [hunk 1]
```diff
@@ -130,6 +145,8 @@ private function getOracleMysqlVersionNumber(string $versionString): string
 
         if ($majorVersion === '5' && $minorVersion === '7') {
             $patchVersion ??= '9';
+        } else {
+            $patchVersion ??= '0';
```
Comment (wmouwen) [inline src/Driver/AbstractMySQLDriver.php:149]: This fallback is added as [`version_compare()`][version-compare] doesn't like trailing dots.

```php
php > var_dump(version_compare('8.4.0', '8.4', '>='));
bool(true)

php > var_dump(version_compare('8.4.', '8.4', '>='));
bool(false)

php > var_dump(version_compare('8.4', '8.4', '>='));
bool(true)

php > var_dump(version_compare('8.4', '8.4.0', '>='));
bool(false)
```

[version-compare]: https://www.php.net/manual/en/function.version-compare.php

Comment (derrabus) [reply src/Driver/AbstractMySQLDriver.php:149]: Where's the trailing dot coming from anyway?

Comment (wmouwen) [reply src/Driver/AbstractMySQLDriver.php:149]: The line above the if-statement defines a nullable string and uses that in a concatenation in this method's return value.

```php
$majorVersion = $versionParts['major'];
$minorVersion = $versionParts['minor'] ?? 0;
$patchVersion = $versionParts['patch'] ?? null;

// ...

return $majorVersion . '.' . $minorVersion . '.' . $patchVersion;
```

File: src/Driver/AbstractMySQLDriver.php [hunk 2]
```diff
@@ -64,6 +65,20 @@ public function createDatabasePlatformForVersion($version)
             }
         } else {
             $oracleMysqlVersion = $this->getOracleMysqlVersionNumber($version);
+
+            if (version_compare($oracleMysqlVersion, '8.4.', '>=')) {
```
Comment (derrabus) [inline src/Driver/AbstractMySQLDriver.php]:
```suggestion
            if (version_compare($oracleMysqlVersion, '8.4.0', '>=')) {
```

**Comment by derrabus** [comment]
Can you please rebase your PR, so #6386 is taken into account and we test your PR against MySQL 8.4?

**Comment by wmouwen** [comment]
The current test that fails is due to a reported MySQL bug: https://bugs.mysql.com/bug.php?id=114876

<details><summary>Details</summary>
<p>

```
There was 1 failure:

1) Doctrine\DBAL\Tests\Functional\ExceptionTest::testInvalidUserName
Failed asserting that exception of type "Doctrine\DBAL\Exception\DriverException" matches expected exception "Doctrine\DBAL\Exception\ConnectionException". Message was: "An exception occurred in the driver: Plugin 'mysql_native_password' is not loaded" at
/home/runner/work/dbal/dbal/src/Driver/Mysqli/Exception/ConnectionFailed.php:34
/home/runner/work/dbal/dbal/src/Driver/Mysqli/Driver.php:54
/home/runner/work/dbal/dbal/src/Connection.php:378
/home/runner/work/dbal/dbal/tests/Functional/ExceptionTest.php:340
/home/runner/work/dbal/dbal/tests/Functional/ExceptionTest.php:303
.

FAILURES!
Tests: 4480, Assertions: 6696, Failures: 1, Skipped: 722, Incomplete: 5.
```

</p>
</details>

**Comment by greg0ire** [comment]
That was addressed in #6388, wasn't it?

**Comment by wmouwen** [comment]
> That was addressed in #6388, wasn't it?

Yep, sure is. Seems like IBM fixed their download issues as well. Could someone re-run the checks?

**Review by greg0ire — DISMISSED** [review]
The code looks OK to me. @derrabus do we have a statement somewhere explaining why it's OK to target 3.9.x here?

File: src/Platforms/MySQL84Platform.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                . ' use MySQL84Platform::createReservedKeywordsList() instead.',
+        );
+
+        return Keywords\MySQL84Keywords::class;
+    }
+
+    public function createSelectSQLBuilder(): SelectSQLBuilder
+    {
+        return AbstractPlatform::createSelectSQLBuilder();
+    }
```
Comment (SenseException) [inline src/Platforms/MySQL84Platform.php]: Is this method override really necessary?

Comment (derrabus) [reply src/Platforms/MySQL84Platform.php]: I don't think so. Let's remove it.

**Comment by derrabus** [comment]
> do we have a statement somewhere explaining why it's OK to target 3.9.x here?

Support for newer database platforms is probably okay-ish…? 🫣

**Comment by greg0ire** [comment]
I think so as well, although it cannot quite formulate why we have that exception.

**Comment by wmouwen** [comment]
The docs make it look like MySQL 8.4 is explicitly [supported](https://www.doctrine-project.org/projects/doctrine-dbal/en/3.9/reference/platforms.html#mysql) by DBAL 3.*, however these new keywords not getting quoted can be seen as a bug. So instead of new platform support, you could call it a bugfix?

**Comment by greg0ire** [comment]
Indeed. Plus, Doctrine does not crash with `PlatformNotSupportedYet` or similar upon using 8.4 right now, so it could definitely be seen as a fix.

**Review by greg0ire — APPROVED** [review]

**Comment by derrabus** [comment]
MySQL 8.4 also removed a couple of keywords, see https://dev.mysql.com/doc/refman/8.4/en/keywords.html#keywords-removed-in-current-series. How shall we proceed with those?

**Review by derrabus — APPROVED** [review]

**Comment by wmouwen** [comment]
> MySQL 8.4 also removed a couple of keywords, see https://dev.mysql.com/doc/refman/8.4/en/keywords.html#keywords-removed-in-current-series. How shall we proceed with those?

Those can be removed using `array_diff()` in the [MySQL84Keywords::getKeywords()](https://github.com/doctrine/dbal/blob/08e34891c150be139f7eebbd45bd33d66a1cd948/src/Platforms/Keywords/MySQL84Keywords.php#L35) method. Could do so in this PR, could do so in a new PR as well.

**Comment by derrabus** [comment]
Let's do that in a follow-up. Thanks!

**Pull request merged**